File: cordova-idcheckio/android/IdcheckioSdk.java
package com.ariadnext.idcheckio;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;

import androidx.annotation.Nullable;

import com.ariadnext.idcheckio.sdk.bean.CISContext;
import com.ariadnext.idcheckio.sdk.bean.SdkEnvironment;
import com.ariadnext.idcheckio.sdk.component.Idcheckio;
import com.ariadnext.idcheckio.sdk.component.IdcheckioView;
import com.ariadnext.idcheckio.sdk.interfaces.ErrorMsg;
import com.ariadnext.idcheckio.sdk.interfaces.IdcheckioCallback;
import com.ariadnext.idcheckio.sdk.interfaces.IdcheckioError;
import com.ariadnext.idcheckio.sdk.interfaces.IdcheckioInteraction;
import com.ariadnext.idcheckio.sdk.interfaces.IdcheckioInteractionInterface;
import com.ariadnext.idcheckio.sdk.interfaces.cis.CISType;
import com.ariadnext.idcheckio.sdk.interfaces.result.IdcheckioResult;
import com.ariadnext.idcheckio.sdk.utils.EnumUtils;
import com.ariadnext.idcheckio.sdk.utils.ExtensionUtilsKt;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class IdcheckioSdk extends CordovaPlugin implements IdcheckioCallback, IdcheckioInteractionInterface {
    private CordovaInterface cordovaInterface;
    private CallbackContext callbackContext;
    private static final String PRELOAD = "preload";
    private static final String ACTIVATE = "activate";
    private static final String START = "start";
    private static final String ANALYZE = "analyze";
    private static final String START_ONLINE = "startOnline";
    private static final int START_REQUEST = 5;
    private static final int START_ONLINE_REQUEST = 6;

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
        switch (requestCode){
            case START_REQUEST:
            case START_ONLINE_REQUEST:
                if(resultCode == Activity.RESULT_OK){
                    String result = intent.getExtras().getString("IDCHECKIO_RESULT", "{}");
                    this.callbackContext.success(result);
                } else if(resultCode == Activity.RESULT_CANCELED){
                    String error = intent.getExtras().getString("ERROR_MSG", "{}");
                    this.callbackContext.error(error);
                }
                break;
        }
    }

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        this.cordovaInterface = cordova;
    }

    private void analyze(String params, String side1ToUpload, String side2ToUpload, Boolean isOnline, String cisContext){
        try {
            Uri side1Uri;
            Uri side2Uri = null;
            IdcheckioView.Builder builder = new IdcheckioView.Builder();
            ParameterUtils.parseParameters(builder, params);
            if(side1ToUpload != null && !side1ToUpload.isEmpty()){
                side1Uri = Uri.parse(side1ToUpload);
            } else {
                this.callbackContext.error("SIDE_1_PARSING_ERROR");
                return;
            }
            if(side2ToUpload != null && !side2ToUpload.isEmpty()){
                side2Uri = Uri.parse(side2ToUpload);
            }
            JSONObject cis = new JSONObject(cisContext);
            String cisTypeString = cis.optString("cisType");
            CISType cisType = null;
            if (!cisTypeString.isEmpty()) {
                cisType = CISType.valueOf(cisTypeString);
            }
            //Start in a custom thread
            Idcheckio.analyze(this.cordovaInterface.getActivity(),
                    builder.captureParams(),
                    side1Uri,
                    side2Uri,
                    this,
                    isOnline,
                    new CISContext(cis.optString("folderUid"),
                            cis.optString("referenceTaskUid"),
                            cis.optString("referenceDocUid"),
                            cisType,
                            cis.optBoolean("biometricConsent")));
        } catch (JSONException ex){
            this.callbackContext.error(ExtensionUtilsKt.toJson(new ErrorMsg(IdcheckioError.INCOMPATIBLE_PARAMETERS, null, ex.getMessage())));
        }
    }

    @Override
    public boolean execute(String action, String rawArgs, CallbackContext callbackContext) {
        try {
            this.callbackContext = callbackContext;
            JSONArray array = new JSONArray(rawArgs);
            switch (action){
                case PRELOAD :
                    Idcheckio.preload(this.cordovaInterface.getActivity(), array.getBoolean(0));
                    this.callbackContext.success();
                    break;
                case ACTIVATE :
                    String environment = array.getString(4);
                    if(!EnumUtils.isValidEnum(SdkEnvironment.class, environment)){
                        callbackContext.error("Wrong SdkEnvironment value.");
                    }
                    Idcheckio.activate(array.getString(0), this, this.cordovaInterface.getActivity(),
                            array.getBoolean(1), array.getBoolean(2), array.getBoolean(3), SdkEnvironment.valueOf(environment));
                    break;
                case ANALYZE :
                    this.analyze(array.get(0).toString(), array.getString(1), array.getString(2), array.getBoolean(3), array.get(4).toString());
                    break;
                case START :
                    Intent intent = new Intent(this.cordovaInterface.getActivity(), IdcheckioActivity.class);
                    intent.putExtra("PARAMS", array.get(0).toString());
                    intent.putExtra("ACTION", START);
                    this.cordovaInterface.startActivityForResult(this, intent, START_REQUEST);
                    break;
                case START_ONLINE :
                    Intent intentOnline = new Intent(this.cordovaInterface.getActivity(), IdcheckioActivity.class);
                    intentOnline.putExtra("PARAMS", array.get(0).toString());
                    intentOnline.putExtra("CIS", array.get(1).toString());
                    intentOnline.putExtra("ACTION", START_ONLINE);
                    this.cordovaInterface.startActivityForResult(this, intentOnline, START_ONLINE_REQUEST);
                    break;
                default:
                    return false;
            }
        } catch (JSONException ex){
            this.callbackContext.error(ex.getMessage());
            return true;
        }
        return true;
    }

    @Override
    public void onInitEnd(boolean success, @Nullable ErrorMsg errorMsg) {
        if(success){
            this.callbackContext.success();
        } else {
            this.callbackContext.error(ExtensionUtilsKt.toJson(errorMsg));
        }
    }

    @Override
    public void onIdcheckioInteraction(IdcheckioInteraction idcheckioInteraction, Object data) {
        switch (idcheckioInteraction) {
            case RESULT:
                this.callbackContext.success(ExtensionUtilsKt.toJson((IdcheckioResult) data));
                break;
            case ERROR:
                this.callbackContext.error(ExtensionUtilsKt.toJson((ErrorMsg) data));
                break;
            default:
                break;
        }
    }
}

File: cordova-idcheckio/android/IdcheckioSdk.kt
package com.ariadnext.idcheckio

import android.app.Activity
import android.content.Intent
import android.net.Uri
import com.ariadnext.idcheckio.sdk.bean.OnlineContext
import com.ariadnext.idcheckio.sdk.bean.SdkEnvironment
import com.ariadnext.idcheckio.sdk.component.Idcheckio
import com.ariadnext.idcheckio.sdk.component.IdcheckioView
import com.ariadnext.idcheckio.sdk.interfaces.*
import com.ariadnext.idcheckio.sdk.interfaces.result.IdcheckioResult
import com.ariadnext.idcheckio.sdk.utils.toJson
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaInterface
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.CordovaWebView
import org.json.JSONArray
import org.json.JSONException
import java.io.FileNotFoundException

class IdcheckioSdk : CordovaPlugin(), IdcheckioCallback, IdcheckioInteractionInterface {
    private var cordovaInterface: CordovaInterface? = null
    private var callbackContext: CallbackContext? = null
    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent) {
        when (requestCode) {
            START_REQUEST, START_ONLINE_REQUEST -> if (resultCode == Activity.RESULT_OK) {
                val result = intent.extras!!.getString("IDCHECKIO_RESULT", "{}")
                callbackContext!!.success(result)
            } else if (resultCode == Activity.RESULT_CANCELED) {
                val error = intent.extras!!.getString("ERROR_MSG", "{}")
                callbackContext!!.error(error)
            }
        }
    }

    override fun initialize(cordova: CordovaInterface, webView: CordovaWebView) {
        super.initialize(cordova, webView)
        cordovaInterface = cordova
    }

    private fun analyze(params: String, side1ToUpload: String?, side2ToUpload: String?, isOnline: Boolean, onlineContextJson : String?) {
        try {
            val onlineContext = OnlineContext.createFrom(onlineContextJson?.takeIf { it.isNotEmpty() && it != "null" } ?: "{}")
            val builder = IdcheckioView.Builder()
            ParameterUtils.parseParameters(builder, params)
            val side1Uri: Uri = if (side1ToUpload != null && side1ToUpload.isNotEmpty()) {
                Uri.parse(side1ToUpload)
            } else {
                callbackContext!!.error("SIDE_1_PARSING_ERROR")
                return
            }
            var side2Uri: Uri? = null
            if (side2ToUpload != null && side2ToUpload.isNotEmpty()) {
                side2Uri = Uri.parse(side2ToUpload)
            }
            //Start in a custom thread
            Idcheckio.analyze(cordovaInterface!!.activity, this, builder.captureParams(), side1Uri, side2Uri, isOnline, onlineContext)
        } catch (ex: JSONException) {
            callbackContext!!.error(ErrorMsg(IdcheckioError.INCOMPATIBLE_PARAMETERS, null, ex.message).toJson())
        } catch (ex: FileNotFoundException) {
            callbackContext!!.error(ErrorMsg(IdcheckioError.UNSUPPORTED_FILE_EXTENSION, null, ex.message).toJson())
        }
    }

    override fun execute(action: String, rawArgs: String, callbackContext: CallbackContext): Boolean {
        try {
            this.callbackContext = callbackContext
            val array = JSONArray(rawArgs)
            when (action) {
                PRELOAD -> {
                    Idcheckio.preload(cordovaInterface!!.activity, array.getBoolean(0))
                    this.callbackContext!!.success()
                }
                ACTIVATE -> {
                    val environment = array.getString(3)
                    Idcheckio.activate(array.getString(0), this, cordovaInterface!!.activity, array.getBoolean(1), array.getBoolean(2),
                        SdkEnvironment.valueOf(environment))
                }
                ANALYZE -> analyze(array[0].toString(), array.getString(1), array.getString(2), array.getBoolean(3), array[4].toString())
                START -> {
                    val intent = Intent(cordovaInterface!!.activity, IdcheckioActivity::class.java)
                    intent.putExtra("PARAMS", array[0].toString())
                    intent.putExtra("ACTION", START)
                    cordovaInterface!!.startActivityForResult(this, intent, START_REQUEST)
                }
                START_ONLINE -> {
                    val intentOnline = Intent(cordovaInterface!!.activity, IdcheckioActivity::class.java)
                    intentOnline.putExtra("PARAMS", array[0].toString())
                    intentOnline.putExtra("ONLINE", array[1].toString())
                    intentOnline.putExtra("ACTION", START_ONLINE)
                    cordovaInterface?.startActivityForResult(this, intentOnline, START_ONLINE_REQUEST)
                }
                else -> return false
            }
        } catch (ex: JSONException) {
            this.callbackContext!!.error(ex.message)
            return true
        }
        return true
    }

    override fun onInitEnd(success: Boolean, error: ErrorMsg?) {
        if (success) {
            callbackContext?.success()
        } else {
            callbackContext?.error(error?.toJson())
        }
    }

    override fun onIdcheckioInteraction(interaction: IdcheckioInteraction, data: Any?) {
        when (interaction) {
            IdcheckioInteraction.RESULT -> callbackContext?.success((data as IdcheckioResult).toJson())
            IdcheckioInteraction.ERROR -> callbackContext?.error((data as ErrorMsg).toJson())
            else -> { }
        }
    }

    companion object {
        private const val PRELOAD = "preload"
        private const val ACTIVATE = "activate"
        private const val START = "start"
        private const val ANALYZE = "analyze"
        private const val START_ONLINE = "startOnline"
        private const val START_REQUEST = 5
        private const val START_ONLINE_REQUEST = 6
    }
}